python/glow/wgr/linear_model/logistic_udfs.py

```diff
@@ -0,0 +1,322 @@
+             |-- beta : double
+             |-- xtgx: array
+             |    |-- element: double
+             |-- xty: array
+             |    |-- element: double
+
+    Returns:
+        transformed Pandas DataFrame containing the aggregated leave-fold-out rows from XtX and XtY
+            schema (specified by the irls_eqn_struct):
+             |-- header_block: string
+             |-- sample_block: string
+             |-- label: string
+             |-- alpha_name : string
+             |-- header: string
+             |-- sort_key: integer
+             |-- beta : double
+             |-- xtgx: array
+             |    |-- element: double
+             |-- xty: array
+             |    |-- element: double
+    """
+    sum_xtgx = pdf['xtgx'].sum()
+    sum_xty = pdf['xty'].sum()
+    mean_beta = pdf['beta'].mean()
+
+    # Use numpy broadcast to subtract each row from the sum
+    pdf['xtgx'] = list(sum_xtgx - np.vstack(pdf['xtgx'].array))
+    pdf['xty'] = list(sum_xty - pdf['xty'])
+    pdf['beta'] = list((pdf.shape[0] * mean_beta - pdf['beta']) / (pdf.shape[0] - 1))
+
+    return pdf
+
+
+@typechecked
+def solve_irls_eqn(key: Tuple, key_pattern: List[str], pdf: pd.DataFrame, labeldf: pd.DataFrame,
+                   alphas: Dict[str, Float]) -> pd.DataFrame:
+    """
+    This function assembles the matrices XtGX, XtY and initial parameter guess B0 for a particular sample_block
+    (where the contribution of that sample_block has been omitted) and solves the equation
+    B = B0 - [(XtGX + I*alpha)]-1 * XtY for a single alpha value, and returns the coefficient vector B, where B has 1
+    element per header in the block X.
+
+    Args:
+        key : unique key identifying the group of rows emitted by a groupBy statement
+        key_pattern : pattern of columns used in the groupBy statement that emitted this group of rows
+        pdf : starting Pandas DataFrame containing the lists of rows from beta, XtGX, and XtY for block X identified
+        by :key:
+            schema (specified by the irls_eqn_struct):
+             |-- header_block: string
+             |-- sample_block: string
+             |-- label: string
+             |-- alpha_name : string
+             |-- header: string
+             |-- sort_key: integer
+             |-- beta : double
+             |-- xtgx: array
+             |    |-- element: double
+             |-- xty: array
+             |    |-- element: double
+        labeldf : Pandas DataFrame containing label values (i. e., the Y in the normal equation above).
+        alphas : dict of {alphaName : alphaValue} for the alpha values to be used
+
+    Returns:
+        transformed Pandas DataFrame containing the coefficient matrix B
+            schema (specified by the model_struct):
+                 |-- header_block: string
+                 |-- sample_block: string
+                 |-- header: string
+                 |-- sort_key: integer
+                 |-- alphas: array
```

Comment:
These arrays always have exactly one value, right? If so, they should just be singleton types and not arrays.

Reply:
I am trying to reuse `RidgeRegression` schemas where possible and in this case the `model_struct` schema is used for both ridge and logistic models.  This also lets us reuse the `apply_model` and `score_models` functions for both ridge and logistic models without needing to check the input DataFrame schema.  It seemed to me to be worth the awkward Spark code to keep the `model_struct` schema fixed with the array types, but let me know if you disagree.

Reply:
Makes sense. Perhaps a comment here in this regard would help.